### PR TITLE
Refactor sbd loading

### PR DESCRIPTION
**Pull request by nelsonkopliku**
## Description

Sbd configuration gets read in the two major processes of the agent:
- Discovery (cluster)
- Facts Gathering (`sbd_config` gatherer)

This PR reduces some duplication of sbd configuration file loading and in discovery/facts gathering.

Notes:
`utils.FindMatches('(?m)^(\w+)=(\S[^#\s]*)', sbdConfigRaw)` has been superseded by `envparse.Parse(sbdConfigFile)`.
A subtle difference here is that `FindMatches` ignores empty values like `SBD_OPTS=` while `envparse` decodes to empty string.

This means that the cluster discovery would also have `"SBD_OPTS": ""` among the other configs.

~Would this break something? :sweat_smile:~ I tried and it does not break things up. 
~Should we make sure to skip empty entries when parsing?~ I believe we should return sbd config entries even if they're empty, see https://github.com/trento-project/agent/pull/145#issuecomment-1321746186

### CODE REVIEW

**Review by fabriziosestito — CHANGES_REQUESTED** [review]
> Would this break something? sweat_smile

@nelsonkopliku  I think that before considering changing this, this question should be answered by adding automated tests where needed

File: internal/cluster/sbd.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ const (
 type SBD struct {
 	cluster string
 	Devices []*SBDDevice
-	Config  map[string]interface{}
+	Config  map[string]string
```
Comment (fabriziosestito) [inline internal/cluster/sbd.go:28]: nice!

File: internal/cluster/sbd.go [hunk 2]
```diff
@@ -86,23 +100,31 @@ func NewSBD(executor utils.CommandExecutor, cluster, sbdPath, sbdConfigPath stri
...
+		}
+	}()
 
+	conf, err := envparse.Parse(sbdConfigFile)
 	if err != nil {
-		return nil, errors.Wrap(err, "could not read sbd config file")
+		return nil, &SBDConfigLoadingError{
+			Type:  SDBConfigDecodingError,
+			inner: errors.Wrap(err, "could not read sbd config file"),
+		}
```
Comment (fabriziosestito) [inline internal/cluster/sbd.go:124]: in case we could remove the extra entries so the output is the same to what we had with the regex

**Comment by nelsonkopliku** [comment]
>  adding automated tests where needed

I made this change on trento web https://github.com/trento-project/web/pull/996 and let e2e tests tell us if integration does not break :sweat_smile: 

Would that be enough?

**Comment by nelsonkopliku** [comment]
@fabriziosestito made the change you suggested.
Now if an existing entry in sbd config is an empty string, that entry gets removed from the parsed config.

This also means that if we do something like
```yaml
facts:
  - name: sbd_opts
    gatherer: sbd_config
    argument: SBD_OPTS # or any other possibly empty entry
```
We'll get a `FactGatheringError` saying `sbd-config-value-not-found`.
We can start like this, it's good enough I believe, anyway I see a little discrepancy with the fact that the config value was actually there, it was just empty.
It depends on what meaning we want to give to the emptiness of a config entry and if that state is an error or something we'd like to know as a fact in our expectations.

File: internal/cluster/sbd.go [hunk 3]
```diff
@@ -86,23 +100,37 @@ func NewSBD(executor utils.CommandExecutor, cluster, sbdPath, sbdConfigPath stri
 	return s, nil
 }
 
-func getSBDConfig(sbdConfigPath string) (map[string]interface{}, error) {
-	sbdConfFile, err := os.Open(sbdConfigPath)
+func LoadSbdConfig(sbdConfigPath string) (map[string]string, error) {
+	sbdConfigFile, err := os.Open(sbdConfigPath)
 	if err != nil {
-		return nil, errors.Wrap(err, "could not open sbd config file")
+		return nil, &SBDConfigLoadingError{
```
Comment (arbulu89) [inline internal/cluster/sbd.go:106]: Do we need such a tailored error handling?
Why don't we return how we were returning before?
`errors.Wrap(err, "could not open sbd config file")`
It looks like an overkill to me

File: internal/factsengine/gatherers/sbd.go [hunk 4]
```diff
@@ -72,22 +70,21 @@ func (g *SBDGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.F
 	return facts, nil
 }
 
-func loadSbdConf(sbdConfigPath string) (map[string]string, error) {
-	sbdConfigFile, err := os.Open(sbdConfigPath)
-	if err != nil {
-		return nil, SBDConfigFileError.Wrap(err.Error())
-	}
+func loadSbdConfig(sbdConfigPath string) (map[string]string, error) {
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sbd.go:73]: I find this function a bit useless here.
Why don't we simply use `cluster.LoadSbdConfig` up in the code, and a put `SBDConfigDecodingError` together with the error message.
The result would be the same, and we would have less "boilerplate" code

File: internal/factsengine/gatherers/sbd.go [hunk 5]
```diff
@@ -50,7 +48,7 @@ func (g *SBDGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.F
 	facts := []entities.Fact{}
 	log.Infof("Starting SBD config Facts gathering")
 
-	conf, err := loadSbdConf(g.configFile)
+	conf, err := loadSbdConfig(g.configFile)
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sbd.go:51]: Here is where we should put the error handling in my opinion.
Directly here, one one error, and that's all

**Review by arbulu89 — COMMENTED** [review]
Folks, sorry for arriving late, but I think that the removal of empty values is a bad idea. We are hiding real configuration, that might be useful in the checks for example.
Senging a new field with an empty value shouldn't be a problem for the web, as most probably we don't use that field.

Besides that, I find the error handling way complicated

File: internal/cluster/sbd.go (hunk 2 above)
Comment (nelsonkopliku) [reply internal/cluster/sbd.go]: At the end I returned to my first proposal of not skipping empty values in config and covered in test.

If this is sorted out we need to agree on https://github.com/trento-project/web/pull/996.

File: internal/cluster/sbd.go [hunk 6]
```diff
@@ -86,23 +86,20 @@ func NewSBD(executor utils.CommandExecutor, cluster, sbdPath, sbdConfigPath stri
...
+func LoadSbdConfig(sbdConfigPath string) (map[string]string, error) {
+	sbdConfigFile, err := os.Open(sbdConfigPath)
 	if err != nil {
 		return nil, errors.Wrap(err, "could not open sbd config file")
 	}
 
-	defer sbdConfFile.Close()
-
-	sbdConfigRaw, err := io.ReadAll(sbdConfFile)
+	defer sbdConfigFile.Close()
```
Comment (arbulu89) [inline internal/cluster/sbd.go:95]: We should aim of checking the error output of the `Close` function.
At least to log it

File: internal/cluster/sbd_test.go [hunk 7]
```diff
@@ -278,9 +278,10 @@ func (suite *SbdTestSuite) TestLoadDeviceDataError() {
 }
 
 func (suite *SbdTestSuite) TestGetSBDConfig() {
```
Comment (arbulu89) [inline internal/cluster/sbd_test.go:280]: Let's change the test name to `TestLoadSbdConfig`

**Review by arbulu89 — APPROVED** [review]
@nelsonkopliku Green light from my side.
I commented 2 minor things that don't need an extra review.

File: internal/factsengine/gatherers/sbd_test.go [hunk 8]
```diff
@@ -17,6 +17,14 @@ func TestSBDGathererTestSuite(t *testing.T) {
 	suite.Run(t, new(SBDGathererTestSuite))
 }
 
+func (suite *SBDGathererTestSuite) TestDefaultSBDConfigGagherer() {
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sbd_test.go:20]: As agreed offline, let's remove this test. We accept the coverage reduction in this case, as in fact, the reduction simply comes for the fact that there are less lines of code. This function was not tested beforehand neither

Comment (fabriziosestito) [reply internal/factsengine/gatherers/sbd_test.go:20]: +1, I lifted the threshold to 1%

Comment (nelsonkopliku) [reply internal/factsengine/gatherers/sbd_test.go:20]: Alright, thanks!

File: internal/cluster/sbd.go (hunk 6 above)
Comment (nelsonkopliku) [reply internal/cluster/sbd.go]: changed

**Pull request merged**